# Exploring Hacker News Posts 

## Project Description:

In this project we will be analysing [hacker_news.csv](https://www.kaggle.com/hacker-news/hacker-news-posts?select=HN_posts_year_to_Sep_26_2016.csv) dataset and try to find which type of posts get more comments Ask HN or Show HN and analysise whether posting time affects number of comments on an average.

In [4]:
#opening hacker_news.csv and converting it to list of lists
import csv
fhand = open("hacker_news.csv")
read = csv.reader(fhand)
hn = list(read)

In [5]:
#first 5 rows of the dataset
hn[:6]

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'],
 ['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '8/4/2016 11:52'],
 ['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20'],
 ['11919867',
  'Technology ventures: From Idea to Enterprise',
  'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429',
  '3',
  '1',
  'hswarna',
  '6/17/2016 0:01'],
 ['10301696',
  'Note by Note: The Making of Steinway L1037 (2007)',
  'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0',
 

In [6]:
#extracting 1st row
header = hn[0]
#removing 1st row from hn
hn = hn[1:]
#header
print(header)

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


In [7]:
#verification that 1st row is successfully deleted
hn[:5]

[['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '8/4/2016 11:52'],
 ['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20'],
 ['11919867',
  'Technology ventures: From Idea to Enterprise',
  'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429',
  '3',
  '1',
  'hswarna',
  '6/17/2016 0:01'],
 ['10301696',
  'Note by Note: The Making of Steinway L1037 (2007)',
  'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0',
  '8',
  '2',
  'walterbell',
  '9/30/2015 4:12']]

In [8]:
#filtering data as ask_posts , show_posts , other_posts
ask_posts = []
show_posts = []
other_posts = []
for row in hn:
    title = row[1].lower()
    if title.startswith("show hn"):
        show_posts.append(row)
    elif title.startswith("ask hn"):
        ask_posts.append(row)
    else:
        other_posts.append(row)
print("Number of 'Ask HN' posts : ",len(ask_posts))
print("Number of 'Show HN' posts : ",len(show_posts))
print("Number of 'Other HN' posts : ",len(other_posts))

Number of 'Ask HN' posts :  1744
Number of 'Show HN' posts :  1162
Number of 'Other HN' posts :  17194


In [9]:
#ask_posts list
ask_posts[:5]

[['12296411',
  'Ask HN: How to improve my personal website?',
  '',
  '2',
  '6',
  'ahmedbaracat',
  '8/16/2016 9:55'],
 ['10610020',
  'Ask HN: Am I the only one outraged by Twitter shutting down share counts?',
  '',
  '28',
  '29',
  'tkfx',
  '11/22/2015 13:43'],
 ['11610310',
  'Ask HN: Aby recent changes to CSS that broke mobile?',
  '',
  '1',
  '1',
  'polskibus',
  '5/2/2016 10:14'],
 ['12210105',
  'Ask HN: Looking for Employee #3 How do I do it?',
  '',
  '1',
  '3',
  'sph130',
  '8/2/2016 14:20'],
 ['10394168',
  'Ask HN: Someone offered to buy my browser extension from me. What now?',
  '',
  '28',
  '17',
  'roykolak',
  '10/15/2015 16:38']]

In [10]:
#show_posts list
show_posts[:5]

[['10627194',
  'Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform',
  'https://iot.seeed.cc',
  '26',
  '22',
  'kfihihc',
  '11/25/2015 14:03'],
 ['10646440',
  'Show HN: Something pointless I made',
  'http://dn.ht/picklecat/',
  '747',
  '102',
  'dhotson',
  '11/29/2015 22:46'],
 ['11590768',
  'Show HN: Shanhu.io, a programming playground powered by e8vm',
  'https://shanhu.io',
  '1',
  '1',
  'h8liu',
  '4/28/2016 18:05'],
 ['12178806',
  'Show HN: Webscope  Easy way for web developers to communicate with Clients',
  'http://webscopeapp.com',
  '3',
  '3',
  'fastbrick',
  '7/28/2016 7:11'],
 ['10872799',
  'Show HN: GeoScreenshot  Easily test Geo-IP based web pages',
  'https://www.geoscreenshot.com/',
  '1',
  '9',
  'kpsychwave',
  '1/9/2016 20:45']]

In [11]:
#sum of comments of ask_posts
total_ask_comments = 0
for row in ask_posts:
    comment_count = int(row[4])
    total_ask_comments+=comment_count
#computing average
avg_ask_comments = total_ask_comments/len(ask_posts)
print("Average number of comments for Ask HN posts is : ",avg_ask_comments)

Average number of comments for Ask HN posts is :  14.038417431192661


In [12]:
#sum of comments of show_posts
total_show_comments = 0
for row in show_posts:
    comment_count = int(row[4])
    total_show_comments+=comment_count
#computing average
avg_show_comments = total_show_comments/len(show_posts)
print("Average number of comments for Show HN posts is : ",avg_show_comments)

Average number of comments for Show HN posts is :  10.31669535283993


From the above analysis we can say that Ask HN posts gets more comment on average compared to Show HN posts. The difference between both of them is 
3.7217220783527303.

In [17]:
#Calculate the amount of ask posts created in each hour of the day, 
#along with the number of comments received
import datetime as dt
result_list = []
for row in ask_posts:
    result_list.append([row[6],int(row[4])])
counts_by_hour = {}
comments_by_hour = {}
for row in result_list:
    date_t = dt.datetime.strptime(row[0],"%m/%d/%Y %H:%M")
    hour = dt.datetime.strftime(date_t,"%H")
    counts_by_hour[hour]=counts_by_hour.get(hour,0)+1
    comments_by_hour[hour]=comments_by_hour.get(hour,0)+row[1]
print("Count By Hour :",counts_by_hour)
print("Comments By Hour : ",comments_by_hour)

Count By Hour : {'09': 45, '13': 85, '10': 59, '14': 107, '16': 108, '23': 68, '12': 73, '17': 100, '15': 116, '21': 109, '20': 80, '02': 58, '18': 109, '03': 54, '05': 46, '19': 110, '01': 60, '22': 71, '08': 48, '04': 47, '00': 55, '06': 44, '07': 34, '11': 58}
Comments By Hour :  {'09': 251, '13': 1253, '10': 793, '14': 1416, '16': 1814, '23': 543, '12': 687, '17': 1146, '15': 4477, '21': 1745, '20': 1722, '02': 1381, '18': 1439, '03': 421, '05': 464, '19': 1188, '01': 683, '22': 479, '08': 492, '04': 337, '00': 447, '06': 397, '07': 267, '11': 641}


In [18]:
#calculating average number of comment according to hours
avg_by_hour = []
for key in counts_by_hour:
    avg_by_hour.append([key,comments_by_hour[key]/counts_by_hour[key]])
print(avg_by_hour)

[['09', 5.5777777777777775], ['13', 14.741176470588234], ['10', 13.440677966101696], ['14', 13.233644859813085], ['16', 16.796296296296298], ['23', 7.985294117647059], ['12', 9.41095890410959], ['17', 11.46], ['15', 38.5948275862069], ['21', 16.009174311926607], ['20', 21.525], ['02', 23.810344827586206], ['18', 13.20183486238532], ['03', 7.796296296296297], ['05', 10.08695652173913], ['19', 10.8], ['01', 11.383333333333333], ['22', 6.746478873239437], ['08', 10.25], ['04', 7.170212765957447], ['00', 8.127272727272727], ['06', 9.022727272727273], ['07', 7.852941176470588], ['11', 11.051724137931034]]


In [19]:
#Top 5 Hours for ask post comments
swap_avg_by_hour = []
for row in avg_by_hour:
    swap_avg_by_hour.append([row[1],row[0]])
print(swap_avg_by_hour)

[[5.5777777777777775, '09'], [14.741176470588234, '13'], [13.440677966101696, '10'], [13.233644859813085, '14'], [16.796296296296298, '16'], [7.985294117647059, '23'], [9.41095890410959, '12'], [11.46, '17'], [38.5948275862069, '15'], [16.009174311926607, '21'], [21.525, '20'], [23.810344827586206, '02'], [13.20183486238532, '18'], [7.796296296296297, '03'], [10.08695652173913, '05'], [10.8, '19'], [11.383333333333333, '01'], [6.746478873239437, '22'], [10.25, '08'], [7.170212765957447, '04'], [8.127272727272727, '00'], [9.022727272727273, '06'], [7.852941176470588, '07'], [11.051724137931034, '11']]


In [20]:
sorted_swap = sorted(swap_avg_by_hour,reverse = True)

In [23]:
print("Top 5 Hours for Ask Post comments")
for r in sorted_swap[:5]:
    hr = r[1]
    parsed_hr = dt.datetime.strptime(hr,"%H")
    hour = dt.datetime.strftime(parsed_hr,"%H:%M")
    print("{} -> {:.2f} average comments per post".format(hour,r[0]))

Top 5 Hours for Ask Post comments
15:00 -> 38.59 average comments per post
02:00 -> 23.81 average comments per post
20:00 -> 21.52 average comments per post
16:00 -> 16.80 average comments per post
21:00 -> 16.01 average comments per post


## Result :
So we see that the **Ask Posts** that are made at around 3:00 PM (EST) or 12:30 AM (IST) gets the most number of average comments i.e 38.59.
The second best time will be around 2:00 AM (EST) or 11:30 AM (IST). The expected average comment is 23.81

## Conclusion :

From the above project we can conclude that **Ask Posts** gets more comments compared to **Show Posts** and the best time to post **Ask HN** posts is 12:30 AM (IST). We also calculated the list of top 5 timings with average comments for ask posts so that people can select a time of their convenience and make posts in order to get maximum comments.